pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/RangeMergeOptimizer.java

```diff
@@ -0,0 +1,152 @@
+   */
+  private static List<String> intersectRanges(List<String> range1, List<String> range2) {
+
+    // Build temporary range predicates to parse the string range values.
+    RangePredicate predicate1 = new RangePredicate(DUMMY_STRING, range1);
+    RangePredicate predicate2 = new RangePredicate(DUMMY_STRING, range2);
+
+    String lowerString1 = predicate1.getLowerBoundary();
+    String upperString1 = predicate1.getUpperBoundary();
+
+    long lower1 = (lowerString1.equals(RangePredicate.UNBOUNDED)) ? Long.MIN_VALUE : Long.valueOf(lowerString1);
+    long upper1 = (upperString1.equals(RangePredicate.UNBOUNDED)) ? Long.MAX_VALUE : Long.valueOf(upperString1);
+
+    String lowerString2 = predicate2.getLowerBoundary();
+    String upperString2 = predicate2.getUpperBoundary();
+
+    long lower2 = (lowerString2.equals(RangePredicate.UNBOUNDED)) ? Long.MIN_VALUE : Long.valueOf(lowerString2);
+    long upper2 = (upperString2.equals(RangePredicate.UNBOUNDED)) ? Long.MAX_VALUE : Long.valueOf(upperString2);
+
+    final StringBuilder stringBuilder = new StringBuilder();
+    if (lower1 > lower2) {
+      stringBuilder.append(
+          (predicate1.includeLowerBoundary() ? RangePredicate.LOWER_INCLUSIVE : RangePredicate.LOWER_EXCLUSIVE));
+      stringBuilder.append(lower1);
+    } else if (lower1 < lower2) {
+      stringBuilder.append(
+          (predicate2.includeLowerBoundary() ? RangePredicate.LOWER_INCLUSIVE : RangePredicate.LOWER_EXCLUSIVE));
+      stringBuilder.append(lower2);
+    } else {
+      if (lower1 == Long.MIN_VALUE) { // lower1 == lower2
+        stringBuilder.append(RangePredicate.LOWER_EXCLUSIVE + RangePredicate.UNBOUNDED); // * always has '('
+      } else {
+        stringBuilder.append(
+            (predicate1.includeLowerBoundary() && predicate2.includeLowerBoundary()) ? RangePredicate.LOWER_INCLUSIVE
+                : RangePredicate.LOWER_EXCLUSIVE);
+        stringBuilder.append(lower1);
+      }
+    }
+
+    stringBuilder.append(RangePredicate.DELIMITER);
+    if (upper1 < upper2) {
+      stringBuilder.append(upper1);
+      stringBuilder.append(
+          (predicate1.includeUpperBoundary() ? RangePredicate.UPPER_INCLUSIVE : RangePredicate.UPPER_EXCLUSIVE));
+    } else if (upper2 < upper1) {
+      stringBuilder.append(upper2);
+      stringBuilder.append(
+          (predicate2.includeUpperBoundary() ? RangePredicate.UPPER_INCLUSIVE : RangePredicate.UPPER_EXCLUSIVE));
+    } else {
+      if (upper1 == Long.MAX_VALUE) { // lower1 == lower2
+        stringBuilder.append(RangePredicate.UNBOUNDED + RangePredicate.UPPER_EXCLUSIVE); // * always has ')'
+      } else {
+        stringBuilder.append(lower1);
+        stringBuilder.append(
+            (predicate1.includeUpperBoundary() && predicate2.includeUpperBoundary()) ? RangePredicate.UPPER_INCLUSIVE
+                : RangePredicate.UPPER_EXCLUSIVE);
+      }
+    }
+    return Collections.singletonList(stringBuilder.toString());
```

Comment:
Does this actually work correctly in the case of degenerate ranges (eg. `a < 30 AND 35 < a`)?